Add a GS limit notice in the Global Styles sidebar

BogdanUngureanu opened the pull request:
Just like the other Global Styles limit notices, the new notice is shown only for free and personal plans. The implementation is based on the existing noticed by refactoring the logic we used in the pre-save GS notice.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/1233

## Proposed Changes

* Add a new notice in the Global Styles sidebar based on the `ComplementaryArea/core/edit-site` slot.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Use ETK build from this branch on your Sandbox
* On a free Simple Site, go to site editor and click on the GS button
* Notice that the upgrade nudge is not displayed
* Make a GS change
* Go to Global Styles sidebar and notice that we now display an upgrade nudge on all pages. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

BogdanUngureanu committed:
Add a GS limit notice in the Global Styles Root Screen

Just like the other Global Styles limit notices, the new notice is shown only for free and personal plans. The implementation is based on the existing noticed by refactoring the logic we used in the pre-save GS notice.

BogdanUngureanu commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/screen-root-notice.tsx:
```diff
@@ -0,0 +1,41 @@
+		}
+	}, [ isVisible ] );
+
+	if ( ! isVisible ) {
```
This prevents the notice to be shown when the user doesn't have any custom styles. However, this means it might not be visible enough for users:

- When a user first click on the GS button, the notice is NOT shown.
- They go to a specific menu and make a change.
- They close the menu since they are satisfied with the changes.

In order for them to see the notice, they'll have to click again on the GS panel; but since they don't plan to make any further changes, the's a smaller chance for them to notice that the changes they've made are actually limited (seeing the notice).

BogdanUngureanu replied:
I've left it this way since the notice needs a design anyway and removing that behavior is simple.

dsas replied:
Maybe we'll want to display different messages depending on if you have any GS vs not having any GS anyway 👍

BogdanUngureanu replied:
Fair enough! I'll leave it as it is for now.

BogdanUngureanu commented:
There's a TS error because it fails to recognize `__experimentalGetCurrentGlobalStylesId`, `getEditedEntityRecord`and `__experimentalGetDirtyEntityRecords`.


TypeScript fails me :(. I'll make the adjustments after the code review.

BogdanUngureanu committed:
Moved useGlobalStylesConfig in a js file so we can bypass the TS errors without using ts-ignore.

BogdanUngureanu committed:
Merge branch 'trunk' into add/global-styles-notice

BogdanUngureanu committed:
Add the upgrade nudge using the ComplementaryArea slot.

BogdanUngureanu commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,52 @@
+
+	return (
+		<Fill name="ComplementaryArea/core/edit-site">
+			{ !! isVisible && isGlobalStylesSidebar && (
```
We'll need to do the condition here because if we are doing an early return, the fill will be introduced at the bottom of the page, which means some additional CSS magic needs to be done.

This way, the notice is added at the top of the sidebar.

dsas replied:
This would be super useful as a code comment

BogdanUngureanu committed:
Added comment explaining why we are doing the condition check inside the fill component.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
Does this really need to be a `.tsx` file?

BogdanUngureanu replied:
Personal preference. :) I've seen that more and more folks are migrating to TS. :)

mmtr replied:
Let's solve all the TS issues then and ensure the file doesn't have any `@ts-ignore` comment.

BogdanUngureanu replied:
I've added the ts-ignore because I'm getting some errors even though that shouldn't happen:
```
TS2339: Property 'getActiveComplementaryArea' does not exist on type 'never'.
```

Which is strange... because I don't think I'm doing anything weird to trigger that error. I've had the same issues with `__experimentalGetCurrentGlobalStylesId` functions too, which is why I kept using a JS file for `use-global-styles-config.js`.

mmtr replied:
Where are you getting that error? I tried removing the `ts-ignore` rule and everything looks good according the IDE and the pre-commit hooks.

BogdanUngureanu replied:
That's strange. I'm getting the error in the IDE.

mmtr replied:
Do you get the error if you run `yarn typecheck`? If you don't, you can probably ignore the IDE error – it's maybe misconfigured.

mmtr replied:
Actually, you should run `yarn run tsc --project apps/editing-toolkit`.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,55 @@
+			{ /*
+			We'll need to do the condition here because if we are doing an early return, the fill will be introduced at the bottom of the page, which means some additional CSS magic needs to be done.
+			*/ }
+			{ !! isVisible && isGlobalStylesSidebar && (
```
Isn't `isVisible` already a boolean? If so, you shouldn't need to cast it with `!!`.

BogdanUngureanu replied:
It needed to be typecasted in the use function. This should simplify a bit the condition.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,55 @@
+
+	useEffect( () => {
+		if ( isVisible && isGlobalStylesSidebar ) {
+			recordUpgradeNoticeShow( GLOBAL_STYLES_SIDEBAR );
```
The parameter is used as a context for the event which should be `site-editor` since this is being tracked in the site editor.

If you want to differentiate it from the other event tracked when the pre-save notice shows up, you'd probably want a new event name.

BogdanUngureanu replied:
Added new events and also added a blog_id property. :)

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/use-global-styles-config.js:
```diff
@@ -0,0 +1,32 @@
+			settings: globalStylesRecord?.settings ?? {},
+		};
+
+		// Do not show the notice if the use is trying to save the default styles.
```
This comment should be more generic, since `useGlobalStylesConfig` doesn't know whether `isVisible` is being used for a notice.

Actually, I think this property should be called `globalStylesInUse` or something like that, since that's really the information we're returning here.

BogdanUngureanu replied:
Good idea, done! :)

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/use-global-styles-config.js:
```diff
@@ -0,0 +1,32 @@
+			Object.keys( globalStylesConfig.settings ).length;
+
+		return {
+			globalStylesConfig,
```
`globalStylesConfig` is never used externally, so we can probably avoid returning it here.

mmtr reviewed (commented):
This work well!

It'd be great if the notice wouldn't have the top margin, so it's right at top the sidebar.

I also consider that the current pre-save notice is no longer relevant, so I'd probably get rid of it (and possibly re-use the same copy in the new permanent notice since it looks a bit clearer).

<img width="277" alt="Screenshot 2023-03-30 at 16 00 29" src="https://user-images.githubusercontent.com/1233880/228862117-c286e2ec-20cc-42e5-a3fb-0276a197684e.png">

> This change means that the sidebar will always be taller than the page

IMO, I think we should fix this before shipping it, and ensure that the sidebar doesn't have the scrollbar if it is not needed:


https://user-images.githubusercontent.com/1233880/228861468-ec47240d-8fcf-4d11-a9ac-9a33fdd2de93.mov

BogdanUngureanu committed:
Refactor the implementation based on code review feedback.

- Add dedicated tracking events for the sidebar notice and add a blog_id parameter (which will help to get some stats about the sites that upgrade)
- Rename isVisible property to globalStylesInUse
- Rename utils.ts to tracks-events.ts for better visibility
- Removed a property since from the use function since it's not needed
- Fixed CSS margin issues in the sidebar

BogdanUngureanu committed:
Merge branch 'trunk' into add/global-styles-notice

BogdanUngureanu committed:
Add blog_id property in the tracks events.

BogdanUngureanu committed:
Remove redundant else branch.

BogdanUngureanu commented:
> It'd be great if the notice wouldn't have the top margin, so it's right at top the sidebar.

Fixed it! :) 

> I also consider that the current pre-save notice is no longer relevant, so I'd probably get rid of it (and possibly re-use the same copy in the new permanent notice since it looks a bit clearer).

I think there's still value in showing the notice in the PrePublish panel too. We've seen that upgrade notices in the editor historically don't convert well; one theory we had was that the user might not want to interrupt their work on their site to upgrade to a paid plan.

> IMO, I think we should fix this before shipping it, and ensure that the sidebar doesn't have the scrollbar if it is not needed:

I've tried again to remove the scrollbar but without much luck (since the changes would require at least overriding the existing styling from Gutenberg core - which will make the implementation less robust). 

IMO I think this styling should be handled in Core because the slot we are using (which is a public API) doesn't accommodate multiple UI elements in that section.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,55 @@
+
+const GLOBAL_STYLES_SIDEBAR = 'edit-site/global-styles';
+
+export function GlobalStylesSidebarNotice() {
```
Ideally, the returned value by this function should be typed.

Additionally, is there any reason why the component is not exported as `default` similarly to the other [two](https://github.com/Automattic/wp-calypso/blob/cc032392b4bbac06203b0e7cdcb3392bda7ca7cc/apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/modal.js) [components](https://github.com/Automattic/wp-calypso/blob/cc032392b4bbac06203b0e7cdcb3392bda7ca7cc/apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/notice.js)?

BogdanUngureanu commented:
Looks like the tests are failing :(

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,55 @@
+import { recordUpgradeNoticeSidebarShow, recordUpgradeSidebarNoticeClick } from './tracks-events';
+import { useGlobalStylesConfig } from './use-global-styles-config';
+
+const GLOBAL_STYLES_SIDEBAR = 'edit-site/global-styles';
```
Nitpick (non-blocking): This is only used once now, so we don't really need a reusable const anymore.

BogdanUngureanu replied:
It's a personal preference to use constants instead of magic strings.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,55 @@
+				<div className="interface-complementary-area">
+					<Notice status="warning" isDismissible={ false } className="wpcom-global-styles-notice">
+						{ createInterpolateElement(
+							__( '<a>Upgrade your plan</a> to keep your style changes.', 'full-site-editing' ),
```
I'm unsure about saying "keep" here, since the styles are never lost – they are just not visible on the frontend, so they are technically kept even if you don't upgrade. 

Could we maybe reuse the same wording used in the other notice to say something like `Your changes include customized styles that will only be visible once you <a>upgrade to a Premium plan</a>`?

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/tracks-events.ts:
```diff
@@ -0,0 +1,43 @@
+declare const wpcomGlobalStyles: { upgradeUrl: string; wpcomBlogId: string | null };
+
+/**
+ * Record an event when a user clicks on the notice from the PrePublish panel.
+ */
+export function recordUpgradePrePublishNoticeClick(): void {
```
Nitpick (non-blocking): The site editor uses the concept of saving instead of publishing, so it might be good to stay consistent here.

```suggestion
 * Record an event when a user clicks on the notice from the pre-save panel.
 */
export function recordUpgradePreSaveNoticeClick(): void {
```

BogdanUngureanu committed:
Rename the event functions and change copy

BogdanUngureanu committed:
Add return type and make the component default.

BogdanUngureanu commented:
Looks like the merge is not blocked anymore, but I still have two issues:
The WPCOM_Global_Styles_Test::test_wpcom_block_global_styles_frontend is failing on line 30:
```1) WPCOM_Global_Styles_Test::test_wpcom_block_global_styles_frontend
Failed asserting that null matches expected 'hotpink'.

/var/www/html/wp-content/plugins/editing-toolkit-plugin/wpcom-global-styles/test/class-wpcom-global-styles-test.php:30
```

There's a JS error in the console when you click on a GS menu item:

```
caught TypeError: Cannot read properties of undefined (reading 'find')
    at Object.trackGlobalStylesMenuSelected [as handler] (wpcom-block-editor-global-styles-menu-selected.js:18:1)
    at delegate-event-tracking.js:116:1
    at Array.forEach (<anonymous>)
    at __webpack_modules__../src/wpcom/features/tracking/delegate-event-tracking.js.__webpack_exports__.default (delegate-event-tracking.js:111:1)
    at HTMLDocument.delegateCaptureListener (tracking.js:865:1)
```

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/index.php:
```diff
@@ -60,6 +60,26 @@ function wpcom_should_limit_global_styles( $blog_id = 0 ) {
+/**
+ * Get the WPCOM blog id of the current site for tracking purposes.
+ */
+function wpcom_global_styles_get_wpcom_current_blog_id() {
```
Can we update `wpcom_should_limit_global_styles` to use this new function and avoid having duplicated code?

BogdanUngureanu replied:
I initially thought about it too, however, this function has a different behavior for disconnected Atomic sites: instead of passing the id = 1, it returns null (since that's the expected value for tracking purposes). In the end, I opted for simplicity and duplicated the code.

I guess we could have a condition where it checks if the id returned is `1`. If it's 1, then we can assume that the blogId returned should be passed as `null`. 

I can make the change if you want, but I think it would give double meaning to the blogId variable (in the current state): sometimes is the WPCOM blog id(most of the time), but sometimes is `1`, the local id of the site (for the disconnected sites).

mmtr replied:
Let's leave it for a follow-up then.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/index.php:
```diff
@@ -60,6 +60,26 @@ function wpcom_should_limit_global_styles( $blog_id = 0 ) {
+		}
+	}
+
+		return null;
```
Minor linting issue.
```suggestion
	return null;
```

BogdanUngureanu committed:
Update apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/index.php

Co-authored-by: Miguel Torres <miguel.torres@automattic.com>

mmtr commented:
I found a regression, not sure if that is related with the test failure.

Previously, you could switch between style variations and the GS panel remained open:

https://user-images.githubusercontent.com/1233880/229159564-750e9d82-2ae6-47bd-be3b-d3f356cccea1.mov

However, with these changes, the GS panel is closed automatically after selecting a style variation:


https://user-images.githubusercontent.com/1233880/229159753-7d155807-be87-4a93-a8e0-3faab9537449.mov

BogdanUngureanu commented:
> I found a regression, not sure if that is related with the test failure.

Yeah, it's a known issue with the approach (more like a limitation)... it was already discussed about it somewhere. The problem is that when the slots change, the slot-fill generic component does a forceUpdate on all fills, which means the local state is reset for all fills and the user is moved back to RootScreen. 

This happens in two scenarios, as you've noticed:
- When the notice is shown the first time
- When the notice is removed (e.g. when you revert back to default state, non-gated, like you did)

Sadly I haven't found a solution for this (and I doubt we can do something about it) :(

BogdanUngureanu committed:
Fix the Fill hard reset when the nudge appears and disappears the first time.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,65 @@
+
+	return (
+		<Fill name="ComplementaryArea/core/edit-site">
+			<>
```
It'd be nice to add comment explaining why we need this apparently useless fragment 🙂

BogdanUngureanu replied:
Good point! :)

BogdanUngureanu committed:
Add docblock explaining why we need the fragment.

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx:
```diff
@@ -0,0 +1,71 @@
+										<ExternalLink
+											href={ wpcomGlobalStyles.upgradeUrl }
+											target="_blank"
+											onClick={ () => recordUpgradeSidebarNoticeClick() }
```
Nitpick (non-blocking): The arrow function is not needed, so we could simplify this a little bit.
```suggestion
											onClick={ recordUpgradeSidebarNoticeClick }
```

mmtr commented on apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/notice.js:
```diff
@@ -73,11 +51,7 @@ function GlobalStylesNoticeComponent() {
-									context: 'site-editor',
-								} )
-							}
+							onClick={ () => recordUpgradePreSaveNoticeClick() }
```
Nitpick (non-blocking): The arrow function is not needed, so we could simplify this a little bit.
```suggestion
							onClick={ recordUpgradePreSaveNoticeClick }
```

BogdanUngureanu committed:
Merge branch 'trunk' into add/global-styles-notice

mmtr reviewed (approved):
Nice, it works great now!

I noted that on mobile viewports the notice is partially duplicated but we can address that if needed on a follow-up (I still think we can/should remove the pre-save notice).

<img width="489" alt="Screenshot 2023-04-03 at 11 27 21" src="https://user-images.githubusercontent.com/1233880/229470195-c80a1851-db67-450a-ad05-5c68c9095883.png">

BogdanUngureanu committed:
Update apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/global-style-sidebar-notice.tsx

Co-authored-by: Miguel Torres <miguel.torres@automattic.com>

BogdanUngureanu committed:
Update apps/editing-toolkit/editing-toolkit-plugin/wpcom-global-styles/notice.js

Co-authored-by: Miguel Torres <miguel.torres@automattic.com>

BogdanUngureanu committed:
Changed the background color in the test trying to fix the failure.

BogdanUngureanu committed:
Try to fix the failing test (take 2).

BogdanUngureanu committed:
Remove filter

BogdanUngureanu committed:
Add cache clear.

BogdanUngureanu committed:
Pass the theme directly.

BogdanUngureanu committed:
Try to fix the failing test.

mmtr committed:
Fix unit test
